docs/cpp_style.rst

```diff
@@ -1,38 +1,142 @@
+   *  ``#include <math.h>`` (Use ``#include <cmath>`` instead).
 
 - Exceptions (We are on our way to **remove** all C++ exception usages in Taichi).
 - Prefix member functions with ``m_`` or ``_``.
 - Virtual function call in constructors/destructors.
 - ``NULL`` (Use ``nullptr`` instead).
 - ``using namespace std;`` in the global scope.
 - ``typedef`` (Use ``using`` instead).
+- Misuse of ``&`` (references).
+
+
+Python style
+------------
+
+We generally follow `PEP8 <https://pep8.org>`_ for Python style.
+
+For example:
+
+.. code-block:: python
+
+    import numpy as np
+    from .lang import my_decorator
+
+    class MyClass:
+        ...
+
+    @my_decorator
+    def my_function(mc, value=None):
+        if value is None:
+            value = np.array([2, 3])
+
+        mc.some_value = value
+        if mc.some_obj is not None:
+            mc.some_method()  # some comments
+        elif mc.some_other_obj is not None:
+            # TODO: some todo notes
+            mc.some_other_method()
+        else:
+            raise NotImplementedError('Some error messages')
+
+Identation
+**********
+We always use **4 spaces** for indent. No tabs.
+
+Naming
+******
+- Variable and function names should consist of lowercase words connected by underscores, e.g. ``num_args`` and ``adaptive_arch_select``.
+  * Non-public APIs should additionally start with ``_``, e.g. ``_kernel_impl``. So that end-users won't get confused.
+  * Ideally public APIs should be short and easy-to-memorize, unless it's harming readability, e.g. ``ti.activate`` instead of ``ti.activate_sparse_snode_at_index``.
+  * Variable names like ``i`` or ``a`` should only occur in trivial places, otherwise is inacceptable.
+  * Constant variables should be consisit of words with all latters capitalize, e.g. ``ti.GUI.PRESS``, ``N``.
+- Class and struct names should consist of words with first letters capitalized, e.g. ``KernelTemplateMapper``.
+  * When using abbreviations in CapWords, capitalize all the letters of the abbreviation, e.g. ``ASTTransformer`` instead of ``AstTransformer``.
+- Filenames should consist of lowercase words connected by underscores, e.g. ``ast_checker.py``.
+
+Dos
+***
+- Make good use of default arguments for simplicity.
+- Use Python decorators for functions when appropriate.
+- Use relative import, e.g. ``from .lang import Matrix`` instead of ``from taichi.lang import Matrix``.
+- The operators tend to get scattered across different columns on the screen, and each operator is moved away from its operand and onto the previous line, e.g.:
+
+.. code-block:: python
+
+   income = (gross_wages
+          + taxable_interest
+          + (dividends - qualified_dividends)
+          - ira_deduction
+          - student_loan_interest)
+
+Don'ts
+******
+- Mixed tabs and spaces. (Please always use 4 spaces for indent)
+- Too long lines. (Please make use of local variables to break it down)
```

Comment:
Wouldn't the linter reject this? If so, it might be great to explicitly write down the line length limits of the linter here. (was it 88?)

Reply:
```suggestion
- Lines longer than 88 characters. (Please make use of local variables to break it down)
```